client/main.cpp

```diff
@@ -50,20 +55,42 @@ int main(int argc, char **argv) {
 
     boost::asio::io_context ioc;
     nibaclient::client_session client_session(host, port, ioc, ctx);
+    std::mutex processor_mutex;
+    std::condition_variable processor_cv;
+    bool processed = true;
 
-    std::thread([&ioc, &client_session] {
+    std::thread([&ioc, &client_session, &processor_mutex, &processor_cv, &processed] {
+        // separate io thread so that getline doesn't block our websocket pingpong
         std::string line;
+        nibautil::stopwatch stopwatch;
         while (std::getline(std::cin, line)) {
             if (line == "exit" || line == "quit")
                 break;
+            std::unique_lock<std::mutex> lock(processor_mutex);
+            processor_cv.wait(lock, [&processed] { return processed; });
```

Comment:
What's this mutex for? As far as I understand, the mutex prevents a user from submitting a second request before the first one finishes. Is it because our low-level APIs are asynchronous?

Reply:
Yes that is why the mutex/cv is here. The stdin thread exists to handle user input without blocking `io_context`(so it could respond to websocket ping-pongs), mostly because I couldn't figure out a cross platform way to handle stdin(we also plan to build the test client on windows at some point) using asio. Then there is another network thread to handle networking(`ioc.run()`) ... the network IO are all synchronous on this thread(since we don't need it to be async). So we really just want synchronous processing while also responding to ping-pongs. To implement the delay properly, we must make sure that the 2nd request doesn't come in before the first is completed, otherwise we wouldn't know the correct delay. The mutex also helps us to safely access `client_session`.

Reply:
Regarding the use of condition_variable, looking at this example, I do not understand why the shared variable (processed/ready) is necessary. Shouldn't it be enough just using mutex lock and notify? https://en.cppreference.com/w/cpp/thread/condition_variable

Reply:
if the 2nd request string comes in after the notification of the 1st request completion how would it know to not wait? The flag variable must be there so the condition variable would wait on it being true (if true then don't wait, else wait until true - while waiting also release the mutex).

Reply:
Is the mutex acquired by `std::unique_lock<std::mutex> lock(processor_mutex);` or `std::lock_guard<std::mutex> lock(g_i_mutex);`? 

If a thread acquires the mutex, it means it can do modifications in the critical section. For example, thread A acquires the mutex. Others will fail to acquire it and thus cannot do modifications. If thread B manages to acquire the mutex, it means thread A has finished its work and released the mutex.

Reply:
`std::lock_guard` its a wrapper to unlock mutex when it goes out of scope, `unique_lock` is similar.

Yes. But that doesn't solve the synchronization problem. Input thread can keep on locking the mutex  and send request. There is no gurantee that the network thread will come in right after the request string is sent to it